scaffolding/src/main.rs

```diff
@@ -59,13 +60,36 @@ fn main() {
     env::set_var("TERM", "notty");
     let (tty_r, tty_w) = tty::pty("sh", COLS as u16, ROWS as u16);
 
-    // Handler program output (tty -> screen) on separate thread.
+    // Handle program output (tty -> screen) on separate thread.
     let (tx_out, rx) = mpsc::channel();
     let (tx_key_press, tx_key_release) = (tx_out.clone(), tx_out.clone());
+
+    let pty_open = Arc::new(AtomicBool::new(true));
+    let pty_open_checker = pty_open.clone();
     thread::spawn(move || {
         let output = Output::new(BufReader::new(tty_r));
-        for cmd in output {
-            tx_out.send(cmd.unwrap_or_else(exit_on_io_error)).unwrap();
+        for result in output {
+            match result {
+                Ok(cmd) => {
+                    tx_out.send(cmd).unwrap();
+                },
+                Err(_) => {
```

Comment:
Running from master branch right now this is the stderr/stdout I see:

```
     Running `target/debug/scaffolding-terminal`
Exiting process due to tty error: Input/output error (os error 5)
thread '<unnamed>' panicked at 'GTK may only be used from the main thread.', /home/wayne/.multirust/cargo/registry/src/github.com-88ac128001ac3a9a/gtk-0.0.6/src/rt.rs:104
stack backtrace:
   1:     0x55cb872d74d0 - sys::backtrace::tracing::imp::write::h752705b18648d1caEIv
   2:     0x55cb872da7bf - panicking::default_handler::_$u7b$$u7b$closure$u7d$$u7d$::closure.45061
   3:     0x55cb872da438 - panicking::default_handler::hd7da873c55525eaaRxA
   4:     0x55cb872d0bbc - sys_common::unwind::begin_unwind_inner::h9832d329b70049c9Xwu
   5:     0x55cb87239ca7 - sys_common::unwind::begin_unwind::h15744914004414859651
                        at ../src/libstd/sys/common/unwind/mod.rs:219
   6:     0x55cb8723a17e - rt::main_quit::hbcb5eaef6bf98b384ia
                        at /home/wayne/projects/notty/scaffolding/<std macros>:3
   7:     0x55cb8722b367 - exit_on_io_error::h10467809337542848798
                        at src/main.rs:129
   8:     0x55cb8722b1fc - fn(std..io..error..Error) .> notty..Command $u7b$exit_on_io_error..$LT$notty..Command$GT$$u7d$::fn_pointer_shim.11254::h342fef473c219868
   9:     0x55cb8722b13b - result::Result<T, E>::unwrap_or_else::h13423017310513651362
                        at ../src/libcore/result.rs:658
  10:     0x55cb872204ee - main::_$u7b$$u7b$closure$u7d$$u7d$::closure.10262
                        at src/main.rs:67
  11:     0x55cb8721fe20 - std::thread::Builder::spawn::_$u7b$$u7b$closure$u7d$$u7d$::_$u7b$$u7b$closure$u7d$$u7d$::closure.10251
                        at ../src/libstd/thread/mod.rs:278
  12:     0x55cb8721fdc9 - sys_common::unwind::try::try_fn::h590124606766957929
                        at ../src/libstd/sys/common/unwind/mod.rs:127
  13:     0x55cb872d675b - __rust_try
  14:     0x55cb872d66ed - sys_common::unwind::inner_try::hfb91b5cfe185b795Ztu
  15:     0x55cb8721fd2e - sys_common::unwind::try::h16460161701113814557
                        at ../src/libstd/sys/common/unwind/mod.rs:123
  16:     0x55cb8721fba0 - std::thread::Builder::spawn::_$u7b$$u7b$closure$u7d$$u7d$::closure.10226
                        at ../src/libstd/thread/mod.rs:278
  17:     0x55cb8722084c - boxed::F.FnBox<A>::call_box::h12809181747460897149
                        at ../src/liballoc/boxed.rs:541
  18:     0x55cb872d8e69 - sys::thread::Thread::new::thread_start::h33a79b9c7e674b88Huz
  19:     0x7f146d688283 - start_thread
  20:     0x7f146de8fa4c - clone
  21:                0x0 - <unknown>
```

This is happening because of https://github.com/withoutboats/notty/blob/master/scaffolding/src/main.rs#L67, where `scaffolding` calls into the gtk library from outside the gtk main thread. The gdk library specifically says that develoeprs should only use gtk and gdk from the thread in which `gtk::init()` and `gtk::main()` are called: https://developer.gnome.org/gdk3/stable/gdk3-Threads.html.

In other words, the current behavior of the program is unambiguously the wrong way to terminate the application. Simply breaking from this read loop and letting the thread terminate will not work either because *something* needs to call `gtk::main_quit()`. Hence the `timeout_add` below that periodically reads the value of `pty_open` which is marked as `false` since we take read errors on the pty as indicating that the process on the other side has closed its end.

The specific error, `Input/output error (os error 5)` shows up on Linux systems as a result of attempting to read a closed file descriptor,
* https://github.com/withoutboats/notty/blob/master/src/output/mod.rs#L279
* https://github.com/withoutboats/notty/blob/master/src/output/mod.rs#L245
* https://github.com/withoutboats/notty/blob/master/scaffolding/tty/src/lib.rs#L95

The specific error here is almost certainly `EIO`: 
* https://github.com/rust-lang/libc/blob/master/src/unix/notbsd/mod.rs#L296
Which is documented on the `read` as:
```
       EIO    I/O error.  This will happen for example when the process is in a background process group, tries to read from its controlling terminal, and either it is ignoring or blocking SIGTTIN or its process group is orphaned.  It may also occur when there is a low-level I/O error while reading from a disk
              or tape.
```

Looking at similar pty-handling code for `xterm`:
* https://github.com/tkztmk/xterm/blob/master/ptydata.c#L213-L214

It looks like the `EIO` assumption above is probably correct. As for how `xterm` handles this error condition, that is probably less useful to consider since it isn't a GTK application, but it does appear to be handled very similarly to the way we are doing it in `scaffolding`--calling some kind of cleanup function on EIO.

I also just did a little more investigation into how `gnome-terminal` works and found that it makes use of the `gnome-vte` widget which in turn handles pty interactions using GLib's portable IO channel construct. Given the way that the file descriptor is handed off from `scaffolding` to the  `notty::Output` struct I don't know that using the GLib IO channel will be possible as an immediate "drop-in" replacement without further refactoring.